src/nupic/engine/Output.cpp

```diff
@@ -47,35 +47,67 @@ Output::~Output() noexcept(false) {
   NTA_CHECK(links_.size() == 0) << "Internal error in region deletion, still has links.";
 }
 
+
 // allocate buffer
-// The 'count' argument comes from the impl by calling getNodeOutputElementCount()
-void Output::initialize(size_t count) {
+void Output::initialize() {
   // reinitialization is ok
   // might happen if initial initialization failed with an
   // exception (elsewhere) and was retried.
   if (data_.has_buffer())
     return;
 
-  nodeOutputElementCount_ = count;
-  size_t dataCount;
-  if (isRegionLevel_)
-    dataCount = count;
-  else
-    dataCount = count * region_->getDimensions().getCount();
-  if (dataCount != 0) {
-    if (data_.getType() == NTA_BasicType_SDR && isRegionLevel_) {
-      const Dimensions& dim = region_->getDimensions();
-      if (dim.isDontcare() || dim.isOnes()) 
-        data_.allocateBuffer(dataCount);
-      else
-        data_.allocateBuffer(dim);
+
+  NTA_CHECK(!dim_.isDontcare())
+        << "Output Dimensions cannot be determined for Region "
+        << region_->getName() << "; output " << name_;
+
+  size_t count = dim_.getCount();
+  if (data_.getType() == NTA_BasicType_SDR) {
+      data_.allocateBuffer(dim_);
+  } else {
+    data_.allocateBuffer(count);
+    // Zero the buffer because unitialized outputs can screw up inspectors,
+    // which look at the output before compute(). NPC-60
+    data_.zeroBuffer();
+  }
+}
+
+
+Dimensions Output::determineDimensions() {
+  if (data_.has_buffer())
+    return dim_;
+
+  if (!dim_.isSpecified()) {
+    dim_.clear();
+    // ask the spec how big the buffer is.
+    const std::shared_ptr<Spec>& srcSpec = region_->getSpec();
+    UInt32 count = (UInt32)srcSpec->outputs.getByName(name_).count; 
+    if (count > 0) {
+      dim_.push_back(count);
     } else {
-      data_.allocateBuffer(dataCount);
-      // Zero the buffer because unitialized outputs can screw up inspectors,
-      // which look at the output before compute(). NPC-60
-      data_.zeroBuffer();
+      // ask the region impl what the output dimensions are.
+      dim_ = region_->askImplForOutputDimensions(name_);
+      if (dim_.isUnspecified()) {
```

Comment:
> Special types of Dimensions

These are all defined in Dimensions.hpp.  I did not change any of this.
- isUnspecified - everything starts out with this state. We have not yet looked to seed if it has been configured with a dimension.
- isDontCare means we have confirmed that it was not configured with a dimension but that it can be inherited from someplace else.  For example, if we looked at an input and checked that it was not configured with a dimension we can mark it as isDontCare so that later when we determine the dimension of the connected output we know that it can also assign it to the input. 
- isSpecified means we have a real dimension assigned. If both ends of a link are specified, we must confirm that both ends of the link are consistent.
- isValid means it has been verified to be a valid dimension (no 0's in wrong places etc).
- isOnes means all dimensions are 1's.  This is kind of a special case since there is really only one element.

Now that I look at the descriptions given in Dimensions.hpp, they don't say much.  I will add some more text.

I don't know how one would simplify it other than perhaps remove isOnes() since I did not use that.

Reply:
> I don't know how one would simplify it 

may not be in this PR, but from what I read here: 
- drop isUnspecified, as it's !isSpecified
- drop isValid, as we always want valid, or immediately error
- isOnes indeed seems overspecific, might drop
- we could drop isDontCare and replace it with a special value of the dims vector (empty, all MAX_INTS?) to idicate this option. Up to you which you find easier/clearer to use?